### DISCUSSION TITLE
quantizeColor

### DISCUSSION THREAD

**Discussion by pasu** [zeux/meshoptimizer#293] [Q&A] [answered]
I can not understand the effect of mask in this function, what is the benefit we can gain from this mask. 

"(mask & -(result >> (bytebits - 1)))"
is this operator correct? I think the result of it is always mask. I can not understand its purpose.

Could anyone explain it? Thank you very much.

`static int quantizeColor(float v, int bytebits, int bits)
{
	int result = meshopt_quantizeUnorm(v, bytebits);

	// replicate the top bit into the low significant bits
	const int mask = (1 << (bytebits - bits)) - 1;

	int nres = (result & ~mask) | (mask & -(result >> (bytebits - 1)));
	return nres;
	//(result & ~mask) | (mask & -(result >> (bytebits - 1)));
}`

**Comment by zeux** [comment, marked as answer]
Imagine that we're using 16 bit colors to serialize in the glTF file. As per glTF spec, the only two fixed-point widths are 8 bit and 16 bit, and you have to store the value in 0..1 range, so you can't e.g. use 12 bits for storage. However, 16 bits is usually overkill from the precision perspective - as such, even for high quality color storage something like 12 bits is enough.

It's possible to simply round the bit count to the next multiple of 8 (12 -> 16) and use that for storage. That's what `meshopt_quantizeUnorm` does (when bits is 12, bytebits is 16). However, gltfpack tries to remove extra noise from the lower bits of the color by masking them out, which results in higher compression ratio.

In this example, when bits is 12 and bytebits is 16, we need to discard 4 lower bits. That's what `mask` stores.

It's tempting to just `return result & ~mask`, but this is incorrect since it loses precision at the endpoint - 1.0 becomes something less than 1.0, which is undesirable for colors. Thus, we set the low 4 bits (in this case) to either all 0s or all 1s depending on the most significant bit of result. This is done via `mask & -(result >> (bytebits - 1))` - the `v & -bit` computation uses the fact that negating a single-bit value results in either all 1s or all 0s depending on whether the bit was 1 or 0, so it's a quick way to replicate a bit to all bits. `mask &` then keeps just the lower bits.

**Comment by pasu** [comment]
I got it. Thank you for your explanation.
I also read the 'Learning from data' from your blog.
I try to use meshoptimizer in our production. 
Thank you for this wonderful project.
